File: mobile/lib/models/chat_command.dart
enum ChatCommandScope { user, channel }

ChatCommandScope _scopeFromString(String s) {
  return s == 'channel' ? ChatCommandScope.channel : ChatCommandScope.user;
}

String chatCommandScopeToString(ChatCommandScope s) =>
    s == ChatCommandScope.channel ? 'channel' : 'user';

class ChatCommand {
  final String id;
  final ChatCommandScope scope;
  final String ownerId;
  final String name;
  final String response;
  final String description;

  const ChatCommand({
    required this.id,
    required this.scope,
    required this.ownerId,
    required this.name,
    required this.response,
    required this.description,
  });

  factory ChatCommand.fromJson(Map<String, dynamic> json) {
    return ChatCommand(
      id: json['id'] as String? ?? json['_id'] as String? ?? '',
      scope: _scopeFromString(json['scope'] as String? ?? 'user'),
      ownerId: json['ownerId'] as String? ?? '',
      name: json['name'] as String? ?? '',
      response: json['response'] as String? ?? '',
      description: json['description'] as String? ?? '',
    );
  }
}

class MyChatCommands {
  final List<ChatCommand> user;
  final List<ChatCommand> channel;

  const MyChatCommands({required this.user, required this.channel});

  factory MyChatCommands.fromJson(Map<String, dynamic> json) {
    final user = (json['user'] as List<dynamic>? ?? [])
        .map((e) => ChatCommand.fromJson(e as Map<String, dynamic>))
        .toList();
    final channel = (json['channel'] as List<dynamic>? ?? [])
        .map((e) => ChatCommand.fromJson(e as Map<String, dynamic>))
        .toList();
    return MyChatCommands(user: user, channel: channel);
  }
}

File: backend/chat/src/models/ChatCommand.ts
import mongoose, { Types } from 'mongoose'

export const CHAT_COMMAND_SCOPE_USER = 'user'
export const CHAT_COMMAND_SCOPE_CHANNEL = 'channel'

export type ChatCommandScope = typeof CHAT_COMMAND_SCOPE_USER | typeof CHAT_COMMAND_SCOPE_CHANNEL

export interface IChatCommand {
    _id: Types.ObjectId
    scope: ChatCommandScope
    ownerId: string
    name: string
    response: string
    description: string
    createdAt: Date
}

const chatCommandSchema = new mongoose.Schema(
    {
        scope: {
            type: String,
            required: true,
            enum: [CHAT_COMMAND_SCOPE_USER, CHAT_COMMAND_SCOPE_CHANNEL],
            index: true
        },
        ownerId: {
            type: String,
            required: true,
            maxlength: 36,
            index: true
        },
        name: {
            type: String,
            required: true,
            maxlength: 32,
            lowercase: true,
            trim: true,
            match: /^[a-z0-9_-]+$/
        },
        response: {
            type: String,
            required: true,
            maxlength: 500
        },
        description: {
            type: String,
            maxlength: 120,
            default: ''
        },
        createdAt: { type: Date, default: Date.now }
    },
    { collection: 'chat_commands' }
)

chatCommandSchema.index({ scope: 1, ownerId: 1, name: 1 }, { unique: true })

const ChatCommand = mongoose.model<IChatCommand>('ChatCommand', chatCommandSchema)

export { ChatCommand }
